storage/innobase/row/row0ins.cc

```diff
@@ -2710,14 +2710,16 @@ row_ins_clust_index_entry_low(
 	    && !trx->check_unique_secondary && !trx->check_foreigns
 	    && !trx->dict_operation
 	    && block->page.id().page_no() == index->page
-	    && !index->table->is_native_online_ddl()
-	    && (!dict_table_is_partition(index->table)
-	        || thd_sql_command(trx->mysql_thd) == SQLCOM_INSERT)) {
+	    && !index->table->is_native_online_ddl()) {
 
+		int sql_com= thd_sql_command(trx->mysql_thd);
 		if (!index->table->n_rec_locks
 		    && !index->table->versioned()
 		    && !index->table->is_temporary()
-		    && !index->table->has_spatial_index()) {
+		    && !index->table->has_spatial_index()
+		    && (!dict_table_is_partition(index->table)
+		        || sql_com == SQLCOM_INSERT
+			|| sql_com == SQLCOM_INSERT_SELECT)) {
```

Comment:
We must not call any non-`inline` function `thd_sql_command()` unless absolutely necessary. I think that we just need to add a Boolean predicate:
```c++
static bool thd_sql_is_insert(const THD *thd) noexcept
{
  switch (thd_sql_command(thd)) {
  case SQLCOM_INSERT:
  case SQLCOM_INSERT_SELECT:
    return false;
  default:
    return true;
  }
}
```
and replace `thd_sql_command(trx->mysql_thd) == SQLCOM_INSERT` with `thd_sql_is_insert(trx->mysql_thd)`.

Reply:
On a second thought, maybe this needs to cover `ALTER TABLE…ALGORITHM=COPY`, possibly by checking for `SQLCOM_ALTER_TABLE`. Can you add a test for that?

Reply:
Alter table ..copy algorithm is already handled by the following if condition:
```
if (flags == (BTR_NO_UNDO_LOG_FLAG | BTR_NO_LOCKING_FLAG)
                   && !index->table->n_rec_locks) {

                ut_ad(index->table->skip_alter_undo);
                ut_ad(!entry->is_metadata());
```

Reply:
I see that there is some test coverage in `innodb.alter_copy_bulk`.